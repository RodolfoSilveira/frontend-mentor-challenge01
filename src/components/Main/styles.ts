import styled, { css } from 'styled-components'
import media from 'styled-media-query'

export const Wrapper = styled.main`
  display: flex;
  justify-content: center;

  ${({ theme }) => css`
    > section {
      background: #fff;
      box-shadow: 1px 2px 6px rgba(0, 0, 0, 0.2);
      width: 50%;
      position: relative;
      top: -6.2rem;
      ${media.lessThan('medium')`
        /* screen width is less than 768px (medium) */
        width: 90%;
      `}

      > article {
        padding: 40px;
        border-bottom: 1px solid #000;

        ${media.lessThan('medium')`
          /* screen width is less than 768px (medium) */
          padding: 20px;
        `}

        ${media.between('medium', 'large')`
          /* screen width is between 768px (medium) and 1170px (large) */
          padding: 20px;
        `}

        div.price-in-month-to-views {
          width: 100%;
          display: grid;
          grid-template-areas:
            'pageviews monthview'
            'range range';
          grid-template-columns: 1fr 1fr;
          align-items: center;

          ${media.lessThan('medium')`
            display: grid;
            grid-template-columns: 1fr;
            grid-template-areas: "pageviews" "range" "monthview";
            margin-bottom: 40px;
          `}

          p.page-count {
            text-transform: uppercase;
            grid-area: pageviews;
            ${media.lessThan('medium')`
              justify-self: center;
            `};
          }

          p.month {
            grid-area: monthview;
            justify-self: end;
            ${media.lessThan('medium')`
              justify-self: center;
            `};

            > span.price {
              font-size: 2rem;
              font-weight: ${theme.font.bold};
              color: ${theme.colors.neutral.darkDesaturatedBlue};
            }
          }
        }

        div.discount {
          display: flex;
          justify-content: flex-end;
          align-items: center;

          ${media.lessThan('medium')`
            /* screen width is less than 768px (medium) */
            justify-content: center;
          `}

          p {
            margin: 0 15px;
            ${media.lessThan('medium')`
              /* screen width is less than 768px (medium) */
              display: flex;
              align-items: center;
            `}

            > span {
              color: ${theme.colors.primary.lightRed};
              font-weight: ${theme.font.bold};
              font-size: ${theme.font.sizes.small};
              margin-left: 10px;
              padding: 3px 6px;
              background: ${theme.colors.primary.lightGrayishRed};
              border-radius: 8px;

              ${media.lessThan('medium')`
                /* screen width is less than 768px (medium) */
                padding: 8px 6px;
                border-radius: 20px;
              `}
            }

            span.inner-discount {
              ${media.lessThan('medium')`
                /* screen width is less than 768px (medium) */
                display: none;
              `}

              ${media.between('medium', 'large')`
                /* screen width is between 768px (medium) and 1170px (large) */
                display: none;
              `}
            }
          }
        }
      }

      > footer {
        padding: 20px 40px;
        display: flex;
        justify-content: space-between;
        align-items: center;
        flex-wrap: wrap;

        ${media.lessThan('medium')`
          /* screen width is less than 768px (medium) */
          justify-content: center;
        `}

        ${media.between('medium', 'large')`
          /* screen width is between 768px (medium) and 1170px (large) */
          padding: 20px 20px;
        `}


        ul > li {
          margin: 15px 0;
          display: flex;
          align-items: center;
          ${media.lessThan('medium')`
            /* screen width is less than 768px (medium) */
            justify-content: center;
          `}

          > svg {
            margin-right: 10px;
            color: ${theme.colors.primary.strongCyan};
          }
        }

        button {
          cursor: pointer;
          border: none;
          border-radius: 20px;
          background: ${theme.colors.neutral.darkDesaturatedBlue};
          color: ${theme.colors.neutral.lightGrayishBlue};
          padding: 14px 46px;
        }
      }
    }
  `};
`

export const Ranger = styled.input.attrs({
  type: 'range',
  min: 8,
  max: 36
})`
  width: 100%;
  margin: 40px 0;
  border-radius: 18px;
  height: 10px;

  grid-area: range;

  ${({ theme }) => css`
    -webkit-appearance: none;
    background: ${theme.colors.primary.softCyan};

    ::-webkit-slider-thumb {
      width: 30px;
      -webkit-appearance: none;
      padding: 19px;
      cursor: pointer;
      background: url('img/icon-slider.svg') no-repeat;
      background-position: 50%;
      border-radius: 100%;
      background-color: ${theme.colors.primary.strongCyan};
      border: none;
      position: relative;
      top: -140%;

      &:active {
        background-color: ${theme.colors.primary.softCyan};
      }
    }

    ::-webkit-slider-runnable-track {
      background-color: ${theme.colors.primary.softCyan};
      height: 100%;
      border-radius: 18px;
      -webkit-appearance: none;
    }

    :focus {
      outline: none;
    }

    ::-moz-range-thumb {
      width: 30px;
      -webkit-appearance: none;
      padding: 12px 4px;
      cursor: pointer;
      background: url('img/icon-slider.svg') no-repeat;
      background-position: 50%;
      border-radius: 100%;
      background-color: ${theme.colors.primary.strongCyan};
      border: none;

      &:active {
        background-color: ${theme.colors.primary.softCyan};
      }
    }

    ::-moz-range-progress {
      background: ${theme.colors.primary.softCyan};
      height: 100%;
      border-radius: 18px;
    }

    ::-moz-range-track {
      background-color: #d3d3d3;
      height: 100%;
      border-radius: 18px;
    }

    ::-ms-thumb {
      width: 30px;
      -webkit-appearance: none;
      padding: 12px 4px;
      cursor: pointer;
      background: url('img/icon-slider.svg') no-repeat;
      background-position: 50%;
      border-radius: 100%;
      background-color: ${theme.colors.primary.strongCyan};
      border: none;

      &:active {
        background-color: ${theme.colors.primary.softCyan};
      }
    }

    ::-ms-fill-lower {
      background: ${theme.colors.primary.softCyan};
      height: 100%;
      border-radius: 18px;
    }

    ::-ms-fill-upper {
      background-color: #d3d3d3;
      height: 100%;
      border-radius: 18px;
    }
  `};
`
