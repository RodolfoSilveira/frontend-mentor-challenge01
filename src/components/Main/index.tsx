import { useState, useCallback, useEffect } from 'react'
import * as S from './styles'
import { IoIosCheckmark } from 'react-icons/io'
import Switch from 'react-switch'

const Main = () => {
  const [checked, setChecked] = useState<any>(false)
  const [value, setValue] = useState('8')
  const [discount, setDiscount] = useState('')
  const [views, setViews] = useState('')

  const handlePageViews = useCallback(() => {
    switch (value) {
      case '8':
        setViews('10k')
        break
      case '12':
        setViews('50k')
        break
      case '16':
        setViews('100k')
        break
      case '24':
        setViews('500k')
        break
      case '36':
        setViews('1M')
        break
      default:
        break
    }
  }, [value])

  useEffect(() => {
    handlePageViews()
  }, [handlePageViews])

  const promotion = useCallback(() => {
    let result
    switch (value) {
      case '8':
        result = 8 - (8 * 25) / 100
        setDiscount(String(result))
        break
      case '12':
        result = 12 - (12 * 25) / 100
        setDiscount(String(result))
        break
      case '16':
        result = 16 - (16 * 25) / 100
        setDiscount(String(result))
        break
      case '24':
        result = 24 - (24 * 25) / 100
        setDiscount(String(result))
        break
      case '36':
        result = 36 - (36 * 25) / 100
        setDiscount(String(result))
        break
      default:
        break
    }
  }, [value])

  useEffect(() => {
    if (checked) promotion()
  }, [checked, promotion])

  return (
    <S.Wrapper>
      <section>
        <article>
          <div className="price-in-month-to-views">
            <p className="page-count">
              {!!views && <span>{views}</span>} Pageviews
            </p>

            <p className="month">
              {checked
                ? !!discount && (
                    <span className="price">
                      ${Number(discount).toFixed(2)}
                    </span>
                  )
                : !!value && (
                    <span className="price">${Number(value).toFixed(2)} </span>
                  )}
              / month
            </p>

            <S.Ranger
              value={value}
              onChange={(e) => setValue(e.target.value)}
            />
          </div>

          <div className="discount">
            <p>Monthly Billing</p>
            <Switch
              height={22}
              width={40}
              handleDiameter={14}
              checkedIcon={false}
              uncheckedIcon={false}
              onColor="#a5f3eb"
              onChange={() => setChecked((prevState: boolean) => !prevState)}
              checked={Boolean(checked)}
            />
            <p>
              Yearly Billing{' '}
              <span>
                25% <span className="inner-discount">discount</span>
              </span>
            </p>
          </div>
        </article>

        <footer>
          <ul>
            <li>
              <IoIosCheckmark /> Unlimited websites
            </li>
            <li>
              <IoIosCheckmark /> 100% data ownership
            </li>
            <li>
              <IoIosCheckmark /> Email reports
            </li>
          </ul>

          <button type="button">Start my trial</button>
        </footer>
      </section>
    </S.Wrapper>
  )
}

export default Main
