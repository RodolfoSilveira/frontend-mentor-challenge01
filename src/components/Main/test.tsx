import React from 'react'
import { screen, fireEvent } from '@testing-library/react'
import { renderWithTheme } from '../../utils/tests/helpers'
import Main from '.'

describe('<Main />', () => {
  it('should render the input ranger', () => {
    const { container } = renderWithTheme(<Main />)

    expect(screen.getByLabelText(/ranger values/i)).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should calc correct without discount', () => {
    const { getByLabelText } = renderWithTheme(<Main />)

    fireEvent.change(getByLabelText(/ranger values/i), {
      target: {
        value: '8'
      }
    })

    expect(getByLabelText(/ranger result/i)).toContainHTML(
      '<span aria-label="ranger result">10k</span>'
    )

    fireEvent.change(getByLabelText(/ranger values/i), {
      target: {
        value: '12'
      }
    })

    expect(getByLabelText(/ranger result/i)).toContainHTML(
      '<span aria-label="ranger result">50k</span>'
    )

    fireEvent.change(getByLabelText(/ranger values/i), {
      target: {
        value: '16'
      }
    })

    expect(getByLabelText(/ranger result/i)).toContainHTML(
      '<span aria-label="ranger result">100k</span>'
    )

    fireEvent.change(getByLabelText(/ranger values/i), {
      target: {
        value: '24'
      }
    })

    expect(getByLabelText(/ranger result/i)).toContainHTML(
      '<span aria-label="ranger result">500k</span>'
    )

    fireEvent.change(getByLabelText(/ranger values/i), {
      target: {
        value: '36'
      }
    })

    expect(getByLabelText(/ranger result/i)).toContainHTML(
      '<span aria-label="ranger result">1M</span>'
    )
  })
})
