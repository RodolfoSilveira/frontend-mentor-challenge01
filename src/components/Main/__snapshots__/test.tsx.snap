// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> should render the input ranger 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 > section {
  background: #fff;
  box-shadow: 1px 2px 6px rgba(0,0,0,0.2);
  width: 50%;
  position: relative;
  top: -6.2rem;
}

.c0 > section > article {
  padding: 40px;
  border-bottom: 1px solid #000;
}

.c0 > section > article div.price-in-month-to-views {
  width: 100%;
  display: grid;
  grid-template-areas: 'pageviews monthview' 'range range';
  grid-template-columns: 1fr 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > section > article div.price-in-month-to-views p.page-count {
  text-transform: uppercase;
  grid-area: pageviews;
}

.c0 > section > article div.price-in-month-to-views p.month {
  grid-area: monthview;
  justify-self: end;
}

.c0 > section > article div.price-in-month-to-views p.month > span.price {
  font-size: 2rem;
  font-weight: 800;
  color: hsl(227,35%,25%);
}

.c0 > section > article div.discount {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > section > article div.discount p {
  margin: 0 15px;
}

.c0 > section > article div.discount p > span {
  color: hsl(15,100%,70%);
  font-weight: 800;
  font-size: 0.9rem;
  margin-left: 10px;
  padding: 3px 6px;
  background: hsl(14,92%,95%);
  border-radius: 8px;
}

.c0 > section > footer {
  padding: 20px 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 > section > footer ul > li {
  margin: 15px 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > section > footer ul > li > svg {
  margin-right: 10px;
  color: hsl(174,86%,45%);
}

.c0 > section > footer button {
  cursor: pointer;
  border: none;
  border-radius: 20px;
  background: hsl(227,35%,25%);
  color: hsl(224,65%,95%);
  padding: 14px 46px;
}

.c1 {
  width: 100%;
  margin: 40px 0;
  border-radius: 18px;
  height: 10px;
  grid-area: range;
  -webkit-appearance: none;
  background: hsl(174,77%,80%);
}

.c1::-webkit-slider-thumb {
  width: 30px;
  -webkit-appearance: none;
  padding: 19px;
  cursor: pointer;
  background: url('img/icon-slider.svg') no-repeat;
  background-position: 50%;
  border-radius: 100%;
  background-color: hsl(174,86%,45%);
  border: none;
  position: relative;
  top: -140%;
}

.c1::-webkit-slider-thumb:active {
  background-color: hsl(174,77%,80%);
}

.c1::-webkit-slider-runnable-track {
  background-color: hsl(174,77%,80%);
  height: 100%;
  border-radius: 18px;
  -webkit-appearance: none;
}

.c1:focus {
  outline: none;
}

.c1::-moz-range-thumb {
  width: 30px;
  -webkit-appearance: none;
  padding: 12px 4px;
  cursor: pointer;
  background: url('img/icon-slider.svg') no-repeat;
  background-position: 50%;
  border-radius: 100%;
  background-color: hsl(174,86%,45%);
  border: none;
}

.c1::-moz-range-thumb:active {
  background-color: hsl(174,77%,80%);
}

.c1::-moz-range-progress {
  background: hsl(174,77%,80%);
  height: 100%;
  border-radius: 18px;
}

.c1::-moz-range-track {
  background-color: #d3d3d3;
  height: 100%;
  border-radius: 18px;
}

.c1::-ms-thumb {
  width: 30px;
  -webkit-appearance: none;
  padding: 12px 4px;
  cursor: pointer;
  background: url('img/icon-slider.svg') no-repeat;
  background-position: 50%;
  border-radius: 100%;
  background-color: hsl(174,86%,45%);
  border: none;
}

.c1::-ms-thumb:active {
  background-color: hsl(174,77%,80%);
}

.c1::-ms-fill-lower {
  background: hsl(174,77%,80%);
  height: 100%;
  border-radius: 18px;
}

.c1::-ms-fill-upper {
  background-color: #d3d3d3;
  height: 100%;
  border-radius: 18px;
}

@media (max-width:768px) {
  .c0 > section {
    width: 90%;
  }
}

@media (max-width:768px) {
  .c0 > section > article {
    padding: 20px;
  }
}

@media (min-width:768px) and (max-width:1170px) {
  .c0 > section > article {
    padding: 20px;
  }
}

@media (max-width:768px) {
  .c0 > section > article div.price-in-month-to-views {
    display: grid;
    grid-template-columns: 1fr;
    grid-template-areas: "pageviews" "range" "monthview";
    margin-bottom: 40px;
  }
}

@media (max-width:768px) {
  .c0 > section > article div.price-in-month-to-views p.page-count {
    justify-self: center;
  }
}

@media (max-width:768px) {
  .c0 > section > article div.price-in-month-to-views p.month {
    justify-self: center;
  }
}

@media (max-width:768px) {
  .c0 > section > article div.discount {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

@media (max-width:768px) {
  .c0 > section > article div.discount p {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }
}

@media (max-width:768px) {
  .c0 > section > article div.discount p > span {
    padding: 8px 6px;
    border-radius: 20px;
  }
}

@media (max-width:768px) {
  .c0 > section > article div.discount p span.inner-discount {
    display: none;
  }
}

@media (min-width:768px) and (max-width:1170px) {
  .c0 > section > article div.discount p span.inner-discount {
    display: none;
  }
}

@media (max-width:768px) {
  .c0 > section > footer {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

@media (min-width:768px) and (max-width:1170px) {
  .c0 > section > footer {
    padding: 20px 20px;
  }
}

@media (max-width:768px) {
  .c0 > section > footer ul > li {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

<main
  class="c0"
>
  <section>
    <article>
      <div
        class="price-in-month-to-views"
      >
        <p
          class="page-count"
        >
          <span
            aria-label="ranger result"
          >
            10k
          </span>
           
          Pageviews
        </p>
        <p
          class="month"
        >
          <span
            class="price"
          >
            $
            8.00
             
          </span>
          / month
        </p>
        <input
          aria-label="ranger values"
          class="c1"
          max="36"
          min="8"
          type="range"
          value="8"
        />
      </div>
      <div
        class="discount"
      >
        <p>
          Monthly Billing
        </p>
        <div
          style="position: relative; display: inline-block; text-align: left; opacity: 1; direction: ltr; border-radius: 11px; transition: opacity 0.25s; user-select: none;"
        >
          <div
            class="react-switch-bg"
            style="height: 22px; width: 40px; margin: 0px; position: relative; background: rgb(136, 136, 136); border-radius: 11px; cursor: pointer; transition: background 0.25s;"
          />
          <div
            class="react-switch-handle"
            style="height: 14px; width: 14px; background: rgb(255, 255, 255); display: inline-block; cursor: pointer; border-radius: 50%; position: absolute; transform: translateX(4px); top: 4px; outline: 0; border: 0px; transition: background-color 0.25s, transform 0.25s, box-shadow 0.15s;"
          />
          <input
            aria-checked="false"
            role="switch"
            style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px;"
            type="checkbox"
          />
        </div>
        <p>
          Yearly Billing
           
          <span>
            25% 
            <span
              class="inner-discount"
            >
              discount
            </span>
          </span>
        </p>
      </div>
    </article>
    <footer>
      <ul>
        <li>
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M362.6 192.9L345 174.8c-.7-.8-1.8-1.2-2.8-1.2-1.1 0-2.1.4-2.8 1.2l-122 122.9-44.4-44.4c-.8-.8-1.8-1.2-2.8-1.2-1 0-2 .4-2.8 1.2l-17.8 17.8c-1.6 1.6-1.6 4.1 0 5.7l56 56c3.6 3.6 8 5.7 11.7 5.7 5.3 0 9.9-3.9 11.6-5.5h.1l133.7-134.4c1.4-1.7 1.4-4.2-.1-5.7z"
            />
          </svg>
           Unlimited websites
        </li>
        <li>
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M362.6 192.9L345 174.8c-.7-.8-1.8-1.2-2.8-1.2-1.1 0-2.1.4-2.8 1.2l-122 122.9-44.4-44.4c-.8-.8-1.8-1.2-2.8-1.2-1 0-2 .4-2.8 1.2l-17.8 17.8c-1.6 1.6-1.6 4.1 0 5.7l56 56c3.6 3.6 8 5.7 11.7 5.7 5.3 0 9.9-3.9 11.6-5.5h.1l133.7-134.4c1.4-1.7 1.4-4.2-.1-5.7z"
            />
          </svg>
           100% data ownership
        </li>
        <li>
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M362.6 192.9L345 174.8c-.7-.8-1.8-1.2-2.8-1.2-1.1 0-2.1.4-2.8 1.2l-122 122.9-44.4-44.4c-.8-.8-1.8-1.2-2.8-1.2-1 0-2 .4-2.8 1.2l-17.8 17.8c-1.6 1.6-1.6 4.1 0 5.7l56 56c3.6 3.6 8 5.7 11.7 5.7 5.3 0 9.9-3.9 11.6-5.5h.1l133.7-134.4c1.4-1.7 1.4-4.2-.1-5.7z"
            />
          </svg>
           Email reports
        </li>
      </ul>
      <button
        type="button"
      >
        Start my trial
      </button>
    </footer>
  </section>
</main>
`;
